src/replication.c

```diff
@@ -212,6 +216,70 @@ void rebaseReplicationBuffer(long long base_repl_offset) {
     }
 }
 
+/* Replication: Master side - connection peering.
+ * On rdb-channel sync, connection peering is used to keep replication data in 
+ * the backlog until the replica requests PSYNC. Peering happens in two forms, 
+ * if there's an existing backlog at the fork time, the replica is attached to 
+ * the tail, if there is no backlog, the replica will be attached when the 
+ * backlog is created. Replica listening ip and port is used as a unique key 
+ * for the peering. On COB overrun, peering is deleted and the RDB connection 
+ * is dropped.
+ */
+void peerPendingSlaveToBacklogBlock(client* slave) {
+    listNode *ln = NULL;
+    replBufBlock *tail = NULL;
+    sds slave_name = sdsnew(replicationGetSlaveNameGeneric(slave, 0));
+    if (server.repl_backlog == NULL) {
+        createReplicationBacklog();
+    } else {
+        ln = listLast(server.repl_buffer_blocks);
+        tail = ln ? listNodeValue(ln) : NULL;
+        if (tail) {
+            tail->refcount++;
+        }
+    }
+    serverLog(LL_DEBUG, "Peer slave %s %s ", slave_name, tail? "with repl-backlog tail": "repl-backlog is empty");
+    slave->ref_repl_buf_node = tail? ln: NULL;
+    dictAdd(server.pending_slaves, slave_name, slave);
+}
+
+/* Attach pending replicas with new replication backlog head. */
+void peerPendingSlavesToBacklogBlockRetrospect(void) {
+    dictIterator *di;
+    dictEntry *de;
+    listNode *ln = listFirst(server.repl_buffer_blocks);
+    replBufBlock *head = ln ? listNodeValue(ln) : NULL;
+    if (head == NULL) return;
+    /* Update pending slaves to wait on new buffer block */
+    di = dictGetSafeIterator(server.pending_slaves);
+    while((de = dictNext(di)) != NULL) {
+        client* slave = dictGetVal(de);
+        slave->ref_repl_buf_node = ln;
+        head->refcount++;
+        serverLog(LL_DEBUG, "Retrospect peer slave %s", (sds)dictGetKey(de));
+    }
+}
+
+void unpeerPendingSlaveFromBacklogBlock(client* slave) {
+    dictEntry *de;
+    listNode *ln;
+    replBufBlock *o;
+    client *peer_slave;
+    sds slave_name = replicationGetSlaveNameGeneric(slave, 0);
```

Comment:
i don't think i'm happy with the fact we're pairing them based on the IP+Port.
what happens if two replicas connect to the same master simultaneously but not at the exact same time?
i.e. they'll have the same ip and port, but different offset, and they deserve different repl buff nodes.
maybe the PSYNC request should carry the client ID of the other connection? (then we can easily find the buffer pointer by looking up the client)
or maybe we should have the dict key be IP+Port+repf_offset?

note that we need to handle both the case of same IP+Port with different offset, and the case of IP+Port with the same offset. so if we don't tie them to client ID, we need some refcount to avoid removing one from this dict too early.

p.s. we need to make sure we have tests covering all these cases.

Reply:
I agree that IP:Port isn't a good dictionary key. IMO adding offset wont help as well because if two replicas are syncing with the same offset, then we will have issues if one of the rdb connections got closed while the other hasn't asked for psync yet (if this client is used as the value of the dictionary then refcount wont help).
Overall breaking client id is harder, but I still have to figure how will the replica main conn send the rdb-client id. I think we will have to add an extra call to replconf, and a new replication state.